### ISSUE TITLE
Avoid internal ConvertedVal allocation when source and target data types are OTHER

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#16370]
The following code:

```java
var val = val(null, OTHER).eq(val(null, OTHER));
System.out.println(val);
```

Leads to the allocation of two instances of `ConvertedVal` in the predicate:

![image](https://github.com/jOOQ/jOOQ/assets/734593/c9f9a6f0-fa14-4b4f-a3b2-12722665910f)

The internal `Val.convertTo()` utility has this check:

```java
// [#11061] Infer bind value data types if they could not be defined eagerly, mostly from the parser.
//          Cannot use convertTo0() here as long as Param.setValue() is possible (mutable Params)
else if (OTHER.equals(getDataType()))
    return new ConvertedVal<>(this, type);
```

But this only makes sense if `!OTHER.equals(type)`

**Comment by lukaseder** [comment]
This is a prerequisite for:

- https://github.com/jOOQ/jOOQ/issues/16368

**Comment by lukaseder** [comment]
This in itself leads to new regressions. Better not tamper with this logic for now. There's a better fix for #16368
